client/src/Features/MicrosoftLearn/MicrosoftLearnFilterCore.ts

```diff
@@ -0,0 +1,217 @@
+  return new Filter({
+    products: parseProductsFromParams(queryParams),
+    roles: parseAsArray(queryParams.get(FilterType.roles)),
+    levels: parseAsArray(queryParams.get(FilterType.levels)),
+    types: parseAsArray(queryParams.get(FilterType.types)),
+    terms: parseAsArray(queryParams.get(FilterType.terms))
+  });
+
+  function parseProductsFromParams(queryParams: URLSearchParams): string[] {
+    const parents = parseAsArray(queryParams.get(FilterType.products));
+    return [...parents, ...addChildrenProducts(parents, products)];
+  }
+
+  function addChildrenProducts(selectedProducts: string[], products: Map<string, Product>): string[] {
+    return [...products.values()]
+      .filter(({ parentId, id }) => parentId && parentId !== id && selectedProducts.includes(parentId))
+      .map(product => product.id);
+  }
+}
+
+export function loadExpandedProductsFromQueryParams(queryParams: URLSearchParams): string[] {
+  return parseAsArray(queryParams.get('expand'));
+}
+
+export function getUpdatedURIFromSelectedFilters(
+  filters: Filter,
+  expandedProducts: string[],
+  products: Map<string, Product> | undefined
+): string {
+  const newQueryParams = new URLSearchParams();
+  addFilteredProductsToQueryParam(newQueryParams);
+  addFiltersToQueryParam(FilterType.roles, newQueryParams);
+  addFiltersToQueryParam(FilterType.types, newQueryParams);
+  addFiltersToQueryParam(FilterType.levels, newQueryParams);
+  addFiltersToQueryParam(FilterType.terms, newQueryParams);
+  addExpandedProductsToQueryParam(newQueryParams);
+  return newQueryParams.toString();
+
+  function addFilteredProductsToQueryParam(queryParams: URLSearchParams): void {
+    const allSelected = getAllProducts(products, filters[FilterType.products]);
+    if (allSelected.length > 0) {
+      queryParams.append(FilterType.products, allSelected.toString());
+    }
+  }
+
+  function addFiltersToQueryParam(filterName: FilterType, queryParams: URLSearchParams): void {
+    if (filters[filterName].filter(item => item !== '').length > 0) {
+      queryParams.append(filterName, filters[filterName].filter(item => item !== '').toString());
+    }
+  }
+
+  function addExpandedProductsToQueryParam(queryParams: URLSearchParams): void {
+    if (expandedProducts.length > 0) {
+      queryParams.append('expanded', expandedProducts.toString());
+    }
+  }
+
+  function getAllProducts(products: Map<string, Product> | undefined, filteredProducts: string[]): string[] {
+    const getParent = getParentProduct(products);
+    const parents = filteredProducts.filter(product => getParent(product) === '');
+    const childrenToInclude = filteredProducts.filter(
+      product => !parents.includes(product) && !parents.includes(getParent(product))
+    );
+    return [...parents, ...childrenToInclude];
+  }
+}
+
+export const getRegexs = (searchTerm: string): RegExp[] => {
+  searchTerm = searchTerm.replace(/\W/gi, '');
```

Comment:
should it not be this?
```suggestion
  searchTerm = searchTerm.replace(/\w/gi, '');
```